### ISSUE TITLE
500 Internal Server Error

### ISSUE THREAD

**Issue by cameronfletcher** [LykkeCorp/guidelines#13]
I do not think that any API we build should by design return a `500 Internal Server Error` status code. I know I am guilty of this in the past.
Discuss.

**Comment by jchannon** [comment]
In Apollo at the time of writing we only return problem+json contents for
500 in dev mode. I suggest we expand that out to production. Not returning
500 is unrealistic you will always get unhandled exceptions and need to
return 500 which is a valid status code. Obviously we would log the
exception. The key is to get internal clients to parse the problem+json
responses but with unhandled exceptions we need to somehow try not to
expose internal information and also have suitable helpful information
which can be tricky if you have something like TaskCancelledException for
example.

**Comment by pawelros** [comment]
I always considered `500` as "Something wrong has happened in our system and we have no idea why". Just like BSOD but scoped to a single request. I also think that `500` by design means unprofessional design.  It should never be the case that we developers accept such errors like something normal. `500` for me means that I have to fix flow that produced that error ASAP.

**Comment by tarurar** [comment]
Same opinion, 500 Internal Server Error is something unexpected and we shouldn't return it.

**Comment by yreynhout-lykke** [comment]
What makes `500` so special? It's just another status code. Why limit ourselves? Because the majority of the infrastructure returns `500` due to unhandled exceptions? True, the number of cases I'd explicitly return `500` are close to 0. But then I don't feel there's a need for that to be prescriptive advice. It's more a case of trying to match http semantics and status codes as closely as possible instead of tunnelling through `500`.

**Comment by JoeStead** [comment]
The `5xx` status code range suggests something on the server side isn't right. 

Looking at this from the HTTP level, exceptions (handled or not) aren't even a thing. If a user makes a valid request, and the server is unable to process it, it is a server error, and the correct status code is in the `5xx` range.

If the user makes an invalid request, then the server so should return a status code in the `4xx` range, and absolutely not `500`.

Looking at this from a HTTP point of view, it becomes quite simple:

The caller made a request that is bad=4xx
The caller made a request that is good, but something on the server is bad=5xx

If the user can't change _anything_ in the request to fix the error, the correct status code is in the`5xx` range.

**Comment by jchannon** [comment]
1xx: hold on
2xx: here you go
3xx: go away
4xx: you fucked up
5xx: I fucked up



On Mon, Feb 11, 2019 at 2:07 PM Joe Stead <notifications@github.com> wrote:
>
> The 5xx status code range suggests something on the server side isn't
right.
>
> Looking at this from the HTTP level, exceptions (handled or not) aren't
even a thing. If a user makes a valid request, and the server is unable to
process it, it is a server error, and the correct status code is in the 5xx
range.
>
> If the user makes an invalid request, then the server so should return a
status code in the 4xx range, and absolutely not 500.
>
> Looking at this from a HTTP point of view, it becomes quite simple:
>
> The caller made a request that is bad=4xx
> The caller made a request that is good, but something on the server is
bad=5xx
>
> If the user can't change anything in the request to fix the error, the
correct status code is in the5xx range.

**Comment by cameronfletcher** [comment]
To add to this, invalid bearer token scopes (in the context of OAuth 2.0) form part of the request and, as such, should result in a `4xx` response.

**Comment by jchannon** [comment]
Agreed, something we check for in Apollo. For example if there is no family name in the ClaimsPrincipal then we know the client has failed to get the data from the IdSrv and when the token comes in and the claim is missing we issue a 400
